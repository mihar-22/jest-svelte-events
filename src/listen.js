// Only for testing.
export let listeners
export let disposals

beforeEach(() => {
  listeners = {}
  disposals = []
})

afterEach(() => { disposals.forEach(fn => fn()) })

export const listen = (component, event) => {
  if (Array.isArray(event)) {
    event.forEach(e => listen(component, e))
    return
  }

  if (!listeners[component]) {
    listeners[component] = {}
    listeners[component].stack = []
  }

  if (listeners[component][event]) return

  listeners[component][event] = { payloads: [], calls: 0 }

  const off = component.$on(event, e => {
    listeners[component].stack.push(event)
    listeners[component][event].payloads.push(e.detail)
    listeners[component][event].calls += 1
  })

  disposals.push(off)
}

export const getListener = component => {
  const listener = listeners[component]
  if (!listener) {
    throw Error(
      'This component has not been registered and has no listeners, ' +
    'setup the listener before calling the matcher.\n\n' +
    'Eg: listen(component, \'myEvent\')\n\n'
    )
  }
  return listener
}

export const getEventData = (component, event) => {
  const listener = getListener(component)
  const eventData = listener[event]
  if (!eventData) {
    throw Error(
    `There are no listeners for the event \`${event}\` on this component, ` +
    'setup the listener before calling the matcher.\n\n' +
    'Eg: listen(component, \'myEvent\')\n\n'
    )
  }
  return eventData
}
