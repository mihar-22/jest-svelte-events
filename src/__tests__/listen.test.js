import '..'

import { listeners, disposals, listen, getListener, getEventData } from '../listen'
import { render, fireEvent } from '@testing-library/svelte'

import Comp from './fixtures/Comp.svelte'

describe('listen', () => {
  test('listeners and disposals should be defined', () => {
    expect(listeners).toBeDefined()
    expect(disposals).toBeDefined()
  })

  test('listening to component should add it to listeners object', () => {
    const { component } = render(Comp)
    listen(component, 'e1')
    expect(listeners[component]).toMatchSnapshot()
    expect(disposals).toHaveLength(1)
  })

  test('listeners should be destroyed after each test', () => {
    expect(listeners).toEqual({})
    expect(disposals).toEqual([])
  })

  test('calling listen on same event should not make any changes', () => {
    const { component } = render(Comp)
    listen(component, 'e1')
    listeners[component].e1.calls = 1
    listen(component, 'e1')
    expect(listeners[component].e1.calls).toBe(1)
    expect(disposals).toHaveLength(1)
  })

  test('firing event should keep track of calls, payloads and stack', async () => {
    const { component, getByText } = render(Comp)
    // e1
    listen(component, 'e1')
    expect(disposals).toHaveLength(1)
    const e1Button = getByText('e1')
    await fireEvent.click(e1Button)
    expect(listeners[component].e1.payloads).toEqual([null])
    expect(listeners[component].e1.calls).toBe(1)
    await fireEvent.click(e1Button)
    expect(listeners[component].e1.payloads).toEqual([null, null])
    expect(listeners[component].e1.calls).toBe(2)
    // e2
    listen(component, 'e2')
    expect(disposals).toHaveLength(2)
    const e2Button = getByText('e2')
    await fireEvent.click(e2Button)
    expect(listeners[component].e2.payloads).toEqual([50])
    expect(listeners[component].e2.calls).toBe(1)
    // stack
    expect(listeners[component].stack).toEqual(['e1', 'e1', 'e2'])
    expect(listeners[component]).toMatchSnapshot()
  })

  test('can listen to multiple events', async () => {
    const { component, getByText } = render(Comp)
    listen(component, ['e1', 'e2', 'e3'])
    expect(disposals).toHaveLength(3)
    const e1Button = getByText('e1')
    const e2Button = getByText('e2')
    const e3Button = getByText('e3')
    await fireEvent.click(e1Button)
    await fireEvent.click(e2Button)
    await fireEvent.click(e3Button)
    expect(listeners[component]).toMatchSnapshot()
  })

  describe('getListener', () => {
    test('returns listener', () => {
      const { component } = render(Comp)
      listen(component, 'e1')
      expect(getListener(component)).toMatchSnapshot()
    })

    test('throws error when component has not been registered', () => {
      const { component } = render(Comp)
      expect(() => {
        expect(component).toHaveFiredEvent('e1')
      }).toThrowErrorMatchingSnapshot()
    })
  })

  describe('getEventData', () => {
    test('returns event data', () => {
      const { component } = render(Comp)
      listen(component, 'e1')
      expect(getEventData(component, 'e1')).toMatchSnapshot()
    })

    test('throws error when event has no listener registered', () => {
      const { component } = render(Comp)
      listen(component, 'e1')
      expect(() => {
        expect(component).toHaveFiredEvent('e2')
      }).toThrowErrorMatchingSnapshot()
    })
  })
})
