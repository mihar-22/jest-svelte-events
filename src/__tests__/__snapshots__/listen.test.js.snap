// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`listen can listen to multiple events 1`] = `
Object {
  "e1": Object {
    "calls": 1,
    "payloads": Array [
      null,
    ],
  },
  "e2": Object {
    "calls": 1,
    "payloads": Array [
      50,
    ],
  },
  "e3": Object {
    "calls": 1,
    "payloads": Array [
      100,
    ],
  },
  "stack": Array [
    "e1",
    "e2",
    "e3",
  ],
}
`;

exports[`listen firing event should keep track of calls, payloads and stack 1`] = `
Object {
  "e1": Object {
    "calls": 2,
    "payloads": Array [
      null,
      null,
    ],
  },
  "e2": Object {
    "calls": 1,
    "payloads": Array [
      50,
    ],
  },
  "stack": Array [
    "e1",
    "e1",
    "e2",
  ],
}
`;

exports[`listen getEventData returns event data 1`] = `
Object {
  "calls": 0,
  "payloads": Array [],
}
`;

exports[`listen getEventData throws error when event has no listener registered 1`] = `
"There are no listeners for the event \`e2\` on this component, setup the listener before calling the matcher.

Eg: listen(component, 'myEvent')

"
`;

exports[`listen getListener returns listener 1`] = `
Object {
  "e1": Object {
    "calls": 0,
    "payloads": Array [],
  },
  "stack": Array [],
}
`;

exports[`listen getListener throws error when component has not been registered 1`] = `
"This component has not been registered and has no listeners, setup the listener before calling the matcher.

Eg: listen(component, 'myEvent')

"
`;

exports[`listen listening to component should add it to listeners object 1`] = `
Object {
  "e1": Object {
    "calls": 0,
    "payloads": Array [],
  },
  "stack": Array [],
}
`;
